### PR TITLE
Issue #152 - First pass at enabling Proxy playback of logged data.  

### PR DESCRIPTION
**Pull request by jsr38**
Reads at rate TESTRATE packets per second, appending to the relevant deque from [TESTDATABASE] when TESTMODE=1, otherwise functions as before.

### CODE REVIEW

File: Proxy/proxy.py
Comment (kb1lqc) [inline Proxy/proxy.py]: Could simplify to use `proxyConfig.getint()` so it will force an int

Comment (jsr38) [reply Proxy/proxy.py]: Sure thing... 👍

File: Proxy/proxy.py
Comment (kb1lqc) [inline Proxy/proxy.py]: Did you mean `conn`?

File: Proxy/proxy.py
Comment (kb1lqc) [inline Proxy/proxy.py:159]: Wouldn't you want to `fetchall()` and then iterate through all returned?

File: Proxy/proxy.py
Comment (jsr38) [reply Proxy/proxy.py]: See subsequent commit.

File: Proxy/proxy.sample.ini
Comment (kb1lqc) [inline Proxy/proxy.sample.ini]: Change sample configuration file callsign & nodeid to `REPLACEME` since it's standard to need to replace them if you are to use it.

File: Proxy/proxy.py
Comment (jsr38) [reply Proxy/proxy.py:159]: Well, it avoids assuming it'll all fit in memory.  Happy to do either.

File: Proxy/proxy.sample.ini
Comment (jsr38) [reply Proxy/proxy.sample.ini]: Sure thing... 👍

File: Proxy/proxy.py
Comment (kb1lqc) [inline Proxy/proxy.py]: Just noting that we should probably fail gracefully on this with a try/except on config read but I won't hold up the PR for this as I am pretty sure `Proxy` hasn't received attention to this at all anyways.

Comment (jsr38) [reply Proxy/proxy.py]: I saw it done in other places but followed suit on this occasion.  I'm gonna touch the files anyhow so I'll do it for these two config reads.

File: Proxy/proxy.sample.ini
Comment (kb1lqc) [inline Proxy/proxy.sample.ini:14]: Just realized if we can have more than one unit in the test mode then we need to be able to identify which unit is which callsign-nodeid. We did this by using separate sections as shown elsewhere in this INI file. However that could start to get messy... Would something like `TESTCALLSIGN0` and similar be acceptable here?

File: Proxy/proxy.py
Comment (kb1lqc) [inline Proxy/proxy.py]: This needs to be a dictionary with key `"data"` as shown [here](https://github.com/FaradayRF/Faraday-Software/blob/master/Proxy/proxy.py#L71). Causes an error when I run `Telemetry` program with a log file.

**Review by kb1lqc — CHANGES_REQUESTED** [review]
I left some comments, noted which ones were more suggestions than necessary but you absolutely must change the `item` in the test worker to be a dictionary with a key of data for this to work. I tested it in my local branch and it works like a charm once implemented!

**Comment by kb1lqc** [comment]
Here's a quick view of the telemetry application showing test proxy data with NO hardware connected once the item["data"] error was resolved (I locally changed it for my testing):

telemetry-testproxy

File: Proxy/proxy.sample.ini
Comment (jsr38) [reply Proxy/proxy.sample.ini:14]: Can run with that.  Multiple test units needs quite a bit more thought.  I'm slightly confused about why the port is written to the DB and not the callsign or nodeid.  At some later point in time, after a configuration has been changed for example, the log no longer makes sense.  Also, I could imagine that the DB log could be interspersed with packets from multiple nodeids and the only way to differentiate is currently port number.  If the callsign and nodeid were logged a single DB could play back traffic for a number of units.  I'm not sure about the design and why it is currently so that may not be what's desired.

File: Proxy/proxy.py
Comment (jsr38) [reply Proxy/proxy.py]: OK didn't spot that.  Will change.  Thanks.

**Comment by kb1lqc** [comment]
Just checked my telemetry database and noticed an interesting (not wrong) observation. Even though my test proxy and telemetry application are setup as `KB1LQC-10` the BASE64 data as it should be is only showing `KB1LQC-1` for both source and destination stations. This is expected because my log file attached to #152 only recorded my local Faraday radio `KB1LQC-1` sending UART data so this would be correct. I simply realized that the proxy/telemetry callsigns are not updated to nodeid 10 which is correct that it should NOT be but I for some reason was surprised to see this ;)

image

**Comment by jsr38** [comment]
Yes, I did wonder about parsing the content of the message.  In fact, when I tested against Telemetry I incorrectly assumed the failure was due to the content not being parsed properly by the Telemetry side.  I think it was late in the day by then as I didn't look any further at it :-( 

Now that I can see the interpreted packets, does it make sense to parse the packet to append to the 'correct' deque?  I don't know anything about the network and it sounds like although you were surprised, you actually want to be able to force the locally connected callsign-nodeid...

File: Proxy/proxy.py
Comment (kb1lqc) [reply Proxy/proxy.py:159]: Ohh you fetchone() at the end. OK I see how this works. That makes total sense. My bad!

File: Proxy/proxy.sample.ini
Comment (kb1lqc) [reply Proxy/proxy.sample.ini:14]: @jsr38 Yeah this is an area where it would help if @kb1lqd and I documented it a bit better. We do need to spend some time on this type of documentation. The port number is essentially an eight bit number (0-255) that indicates a potential application on Faraday itself written in firmware. For example `port 5` is the telemetry application on Faraday so any data received from port 5 will be related to telemetry. Any 'port 2' data is intended for the command/control application that controls GPIO and other settings. Think of it as the UART COM is an "IP Address" and the `Port` is the one of many ports available on an IP address (i.e. 80 for HTTP, 22 for SSH, etc)... Similar concept

**Comment by kb1lqc** [comment]
Yeah each deque dictionary is organized with a callsign-nodeid key so that probably makes sense to abide by.

**Comment by kb1lqc** [comment]
An observation on variable/function naming schemes. We technically have [guidelines for that now](https://github.com/FaradayRF/Faraday-Software/blob/master/CONTRIBUTING.md#python). @kb1lqd and I were not consistent in this when we started the project but we should enforce it more. Your new code doesn't follow this and I wouldn't hold the old code to that level of scrutiny on this PR (likely another PR) but we should be doing our best to abide by it.

I don't want to bog you down with this formatting stuff, I might update it for you!

File: Proxy/proxy.py
Comment (kb1lqc) [inline Proxy/proxy.py]: Yup and right here getDicts[unit][port] is correctly adding the data to a dictionary of dequeues for each station just like the normal Proxy. Looks good!

**Comment by kb1lqc** [comment]
On second thought proxy needs to just be updated at once for the function names and stuff. Lets really only focus on new code with this unless you would really like to go back and fix this @jsr38. I'll approve but leave it up to you for this.

**Review by kb1lqc — APPROVED** [review]
Looks good. I've noticed a few items that should be noted:
- Does not have ability to "repeat" log file. Makes sense and wasn't actually a requirement but my natural assumption was that it did. I think this is actually good, Do not repeat.
- Variable and function names do not conform to `contributions.md`.
- If you set `LOG=1` and `TESTMODE=1` then you enter playback mode but do not log to a new database. Makes sense but simply noting that someone can attempt to setup the INI file like this.

I've let @jsr38 decide if he'd like to update the proxy contributions to follow the conventions we started following but since Proxy is and old contribution much of the code already written prior to this PR doesn't follow this either.

**Pull request merged**
